### DISCUSSION TITLE
VX3: ElectricalEnergySystemPhotovoltaicStatus plötzlich geändertes Verhalten?

### DISCUSSION THREAD

**Discussion by M4n197** [open3e/open3e#70] [Q&A] [answered]
Hallo allerseits,
vorweg hoffe ich hier jemanden zu finden, der seinen Vitocharge VX3 bereits mit open3e ausliest. 
Wenn diese Frage besser im Viessmann-Forum aufgehoben ist, entschuldige ich mich an dieser Stelle.
Da die Diagnostizierung ohne open3e aber wohl nahezu unmöglich ist, wollte ich es erstmal hier versuchen. 

### Problem
Ich habe eine ungewöhnliche Veränderung beim Auswerten der Leistung der PV-Strings von meinem VX3 festgestellt.
Bisher habe ich über `DID 1690` die Leistungen "ActivePower String 1", "ActivePower String 2" und  "ActivePower String 3" abgefragt und geloggt, wobei "ActivePower String 2" immer 0 ist und die anderen demnach den wirklich installierten PV-Strings entsprechen.
**Seit dem 24.01.2024 liefern mir "ActivePower String 1" und "ActivePower String 3" den gleichen Wert**, die Kurven liegen nun exakt aufeinander. 
Zeitpunkt der "Problementstehung" (~10:48):
![grafik](https://github.com/open3e/open3e/assets/142239255/6bc4e2f3-1d54-4bec-b89c-9b30787643f8)

### Erklärungsversuche
1. Dass zwei verschiedene Strings, die sich in Modulanzahl und Neigungswinkel (~45° & ~13°) unterscheiden plötzlich die gleiche Leistung liefern, ist mMn. nicht plausibel. 
2. Empfangs- und Verarbeitungsfehler schließe ich eher aus, da die CAN-Messages bereits im Rohformat an den entsprechenden Stellen die gleichen HEX-Werte aufweisen. Das Setup für die Auswertung wurde nicht verändert.
3. Leider habe ich noch keine Aufzeichnung zur aktuellen Firmware-Version. Kann die Firmware überhaupt aus der Ferne aktualisiert werden? Da kein Neustart des Systems zu erkennen war, halte ich das für unwahrscheinlich.
4. Evtl. gibt es einen Defekt. Fehler wirft der VX3 jedoch keine. Warnungen gab es in den letzten Monaten ab und an die "_A.106 - 96 Isolationsüberwachung konnte nicht ausgeführt werden._" Dadurch startet der VX3 mit der Produktion verspätet, und lässt gute Sonnenstunden verstreichen. Das klingt nicht sonderlich toll, aber ist wahrscheinlich nicht die Ursache dieses Problems. 

### weitere Schritte
1. Evtl. liefert `1831 : O3EComplexType(12, "PhotovoltaicCurrentStringPower"...`  andere Werte. Das werde ich demnächst einrichten.
2. Die generierte Energie aus `1802 : O3EComplexType(80, "EnergyProductionPhotovoltaic"` sollte ich auch näherungsweise durch die Integration der Leistungen bestimmen können. Nur durch "scharfes Hinsehen" konnte ich jedoch bisher keine Diskrepanz erkennen.
3. Ohne andere Erkenntnisse werde ich bei Gelegenheit je einen String abklemmen und den Effekt beobachten. 

### Ausschnitt zu `DID 1690` aus Open3Edatapoints.py
```
        1690 : O3EComplexType(17, "ElectricalEnergySystemPhotovoltaicStatus", 
                                  [O3EInt16(2, "ActivePower String 1", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 1", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower String 2", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 2", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower String 3", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 3", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower cumulated", scale=1.0, signed=True), O3EInt16(2, "RectivePower cumulated", scale=1.0, signed=True),
                                   O3EInt8(1, "OpMode", scale=1.0, signed=False)]),
```
Hier hat sich mit "RectivePower  ..." statt "ReactivePower  ..." wohl noch ein Tippfehler eingeschlichen? Der Sinn dieses Wertes erschließt sich mir in dem Kontext trotzdem nicht. Bei mir bleibt der auch immer 0.

**Comment by philippoo66** [comment]
Moin @M4n197 !

`Hier hat sich mit "RectivePower ..." statt "ReactivePower ..." wohl noch ein Tippfehler eingeschlichen?`

das ist wahrscheinlich. Kann gut sein, dass ich das verbockt hab.

Eigentlich kann ich mir deine Beobachtung fast nur durch ein SW Update (mit Fehler in der SW) erklären. Ich meine, das wäre aus der Ferne möglich und wird auch gemacht. Ein technischer Defekt ist natürlich auch nicht ausgeschlossen...

Ist es möglich, die Anlage mal neu zu starten (ich weiss, mit den Batterien ist das nicht per Stecker Ziehen getan...)

Wir haben hier definitiv mehrere mit VX3, mal schauen, was die dazu berichten können...

Grüsse!
Phil

**Reply by M4n197** [reply]
Hallo Phil,
danke für die Rückmeldung!

Da das so "seamless" war, wollte ich zunächst nicht dran glauben, aber eine andere Erklärung fällt mir nicht ein.

Einen Neustart werde ich mal probieren, wenn ich nächste Woche vor Ort bin. 
Vom Batterietrennschalter und der dazugehörigen Prozedur habe ich schonmal in der Anleitung gelesen.

Die aktuelle Firmware kann ich aktuell nur der GridBox entnehmen:
![grafik](https://github.com/open3e/open3e/assets/142239255/31567bc6-ec84-4f9b-95df-c4cdb80a3497)
Die "24" sieht verdächtig aus, aber ohne Aufzeichnungen kann ich nur mutmaßen...

Beste Grüße

**Comment by M4n197** [comment]
Eine weitere (subtile) Änderung zeigt sich nur Nachts:  
Vorher waren die Leistungen beider Strings in der Nacht logischerweise immer glatt 0W. 
Neuerdings liefert String1 stattdessen durchgehend "Nadeln" mit 1-3W. Auf String3 tut sich aber nichts.
![grafik](https://github.com/open3e/open3e/assets/142239255/3cfc5f8e-7261-4856-9a5f-f9ec0602229b)

**Reply by AmansRevenger** [reply]
Hi 

Ich habe ein ähnliches Bild mit String 1 und 2 

![Screenshot 2024-02-01 at 08 16 03](https://github.com/open3e/open3e/assets/18306225/deba4956-d318-485b-b715-7c09c443458a)
![Screenshot 2024-02-01 at 08 16 17](https://github.com/open3e/open3e/assets/18306225/d23a9ff0-d281-44f7-a75f-5fa94dad3def)
![Screenshot 2024-02-01 at 08 16 30](https://github.com/open3e/open3e/assets/18306225/874de24b-25cb-4bd9-aea5-23c228d87e1d)

String 3 ist der einzige der "richtig" reported (gecheckt mit der App ViCare) und auch sonst stimmen die Daten

Der Setup ist allerdings , das ich per Laptop (!) mit Xubuntu mit dem USB CAN Adapter an der Vitocal250 hänge und von dort die VX3 werte mit einer angepassten Open3eClass.py auslese. Es ist ein temporärer Setup da ich erstmal überhaupt wissen wollte ob es funktioniert.

Hier die angepasste , aktuelle (ish) Open3eClass.py.

![telegram-cloud-photo-size-2-5309959491174913341-y](https://github.com/open3e/open3e/assets/18306225/40394742-49b7-4db6-9070-96692b378b0d)


Meine Sensoren in HA sind wie folgt (String 1 bis 3 austauschbar)

```
  - unique_id: PVProductionString1FromCAN
    name: PV Production on String1 from CAN
    device: *vx3
    state_topic: open3e/vx3/1831_PhotovoltaicCurrentStringPower/String1
    device_class: power
    value_template: |-

      {% if value_json | float < 65000.0 %}
          {{ value_json | float }}
      {% endif %}
    unit_of_measurement: W
    qos: 0
```

und der aufruf auf dem tatsächlichen Laptop ist

`python3 Open3Eclient.py -c can0 -r 378,379,380,680,1552,1603,1664,1690,1801,1802,1807,1808,1809,1810,1822,1823,1824,1825,1826,1827,1828,1829,1830,1831,1832,1833,1834,1836,1837,1838,1839,1840,2240 -dev vx3 -v -m localhost:1883:open3e/vx3 -mfstr "{didNumber}_{didName}" -t 15 `

D.h. ich lese DID 1831 aus, richtig? Ich habe es noch nie geschafft, das das detectSystems ohne Timeout durchläuft, deshalb kann ich hier nur bedingt Aussagen treffen

**Reply by M4n197** [reply]
Hi, 
danke für dein Feedback.

Habe ich das richtig verstanden, dass du nur einen String (String 3) hast?
Negative Leistung auf String2 wäre seltsam und erst recht, wenn gar keiner angeschlossen ist.

Kannst du mal die Ergebnisse von `DID 1831` und `DID 1690` vergleichen?

Welche Firmware-Version ist auf deinem VX3 installiert? In der ViCare-App sehe ich da jedoch auch keine Info. 
Beste Grüße!

**Reply by AmansRevenger** [reply]
Hi

Es ist leider nicht meine Anlage sondern die von meinen Eltern und ich habe nur remote auf den Laptop Zugriff, ich werde bei Gelegenheit nachschauen wegen der Firmware aber das kann Wochen dauern...

Auch deshalb kann ich dir bei den Strings nicht sagen ob etwas verbunden ist.

Jedoch habe ich bei mehreren Tagen Daten keinerlei sinnvolle Werte auf String 1 und 2,  die DIDS muss ich morgen vergleichen

**Reply by philippoo66** [reply]
Moin ihr!

ich hab grad den Hinweis bekommen, dass die String-Werte wohl in umgekehrter Reihenfolge zugeordnet werden:

```
aktuell: vx3 1831 PhotovoltaicCurrentStringPower {"String1": 0.0, "String2": 1171.0, "String3": 1341.0}
am vx3 : vx3 1831 PhotovoltaicCurrentStringPower {"StringC": 0.0, "StringB": 1171.0, "StringA": 1341.0}

aktuell: vx3 1833 PhotovoltaicStringVoltage {"String1": 0.0, "String2": 478.135, "String3": 557.883}
am vx3 : vx3 1833 PhotovoltaicStringVoltage {"StringC": 0.0, "StringB": 478.135, "StringA": 557.883}
```

Könnt ihr das bestätigen? Würde dann ja zumindest erklären, warum bei nur einem String Werte für String "3" angezeigt werden...

**Reply by M4n197** [reply]
Hey, danke für die Info!

Noch nicht wirklich.
Bei `1690` ist es so, dass der mittlere Wert ("String2" / "String B") immer Null liefert (vor und nach dem "Bug"). Die Zuordnung der Strings und Anschlüsse am VX3 prüfe ich mal wenn ich vor Ort bin. Ich werde dann meine Erkenntnis für `1690` hier in analoger Form festhalten.

Morgen werde ich `1831` meiner Auswertung hinzufügen, dann kann ich erstmal die Unterschiede zu `1690` evaluieren. 

Der Anschluss für einen dritten String ist bidirektional und findet stattdessen ggf. für die Batterie Verwendung:
![grafik](https://github.com/open3e/open3e/assets/142239255/aab93102-61de-41b8-9f56-460c29bd5a08)
Bei allen Setups mit Batterie erwarte ich dann auch eine 0 an der gleichen Stelle der Abfrage. 
Bei `1831` wäre das nach neuem Stand "StringC" und würde sich mit dem Ergebnis von @AmansRevenger decken (erstes Bild).

**Reply by rsch90** [reply]
Bei mir sind zwei Strings angeschlossen (1+2) sowie 3 immer 0. habe aber auch eine Batterie angeschlossen. 

Korrektur: String1 ist immer 0 bei mir.

**Reply by M4n197** [reply]
Hallo @rsch90,
welche DID liest du -  `1831` oder `1690`?
Weißt du welche Frimware-Version bei deinem VX3 aktuell installiert ist?


Ich habe nun auch `1831` eingerichtet und erhalte nur auf dem mittleren Wert "String2"/"StringB" (gelb) Ergebnisse, die anderen liefern jeweils Null:
![grafik](https://github.com/open3e/open3e/assets/142239255/8257baa0-0453-450c-9b5f-f8c6c1c32102)
(blau und violett sind die seit dem "Bug" nahezu identischen Werte von DID `1690`)

**Reply by M4n197** [reply, marked as answer]
Ich habe mein Problem gefunden...
![grafik](https://github.com/open3e/open3e/assets/142239255/afb32b4c-c28b-480f-99bb-c1d6e681bac7)
Eines der vier Schütze des Feuerwehrschalters ist aus irgendeinem Grund abgefallen.

Mein "Bug" ist damit eine Kombinationen aus diesem Störfall und der inkorrekten Definition von DID `1690`.
Da ich nur an den String-Leistungen interessiert war, habe ich nie alle Werte von DID `1690` gelesen bzw. aufgezeichnet.
Jetzt ist mir erst aufgefallen, dass die Zeile 
```
O3EInt16(2, "ActivePower cumulated", scale=1.0, signed=True), O3EInt16(2, "RectivePower cumulated", scale=1.0, signed=True),
```
am Anfang stehen müsste. Statt meiner beiden Strings habe ich also bisher fälschlicherweise die Gesamt-PV-Leistung und die Leistung des 2. Strings gelesen.
Korrekt müsste es also wie folgt heißen:
```
1690 : O3EComplexType(17, "ElectricalEnergySystemPhotovoltaicStatus", 
                                   [O3EInt16(2, "ActivePower cumulated", scale=1.0, signed=True), O3EInt16(2, "RectivePower cumulated", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower String 1", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 1", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower String 2", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 2", scale=1.0, signed=True),
                                   O3EInt16(2, "ActivePower String 3", scale=1.0, signed=True), O3EInt16(2, "RectivePower String 3", scale=1.0, signed=True),
                                   O3EInt8(1, "OpMode", scale=1.0, signed=False)]),
```

**Reply by M4n197** [reply]
Wird Zeit, dass ich den Meldekontakt des Feuerwehrschalters noch mit einlese... ^^

**Reply by rsch90** [reply]
@M4n197 Um deine Frage der Vollständigkeit halber noch zu beantworten. Ich nutze 1831, hatte mich aber vertan (war unterwegs) Es ist String 1, der bei mir immer 0 ist.

**Reply by M4n197** [reply]
Hier folgt noch eine kleine Ergänzung bzgl. der erwähnten Warnung `A.106 - 96 Isolationsüberwachung konnte nicht ausgeführt werden.`:
Der Installateur meinte: "Der Fehler ist bekannt und soll mittels Software Update behoben werden." 
In zwei Wochen wird dann ein Viessmann-Servicetechniker vor Ort das Update einspielen.

Sofern jemand ebenfalls diese Warnungen bei seinem VX3 hat, lohnt sich ggf. eine Nachfrage. 
Die Warnung ist leider nur über das Display abrufbar, in der ViCare-App werden sie mir nicht angezeigt.

**Reply by philippoo66** [reply]
du hast doch open3e?! 
```
        263 : O3EList(122, "WarningDtcList", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Warning","Warnings"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        264 : O3EList(124, "WarningDtcHistory", [O3EByteVal(2, "Count"), O3EByteVal(2, "GrandTotal"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Warning","Warnings"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
```
steht sie da nicht drin? ansonsten gibt es ja noch
```
        257 : O3EList(122, "StatusDtcList", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"State","States"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        258 : O3EList(122, "StatusDtcHistory", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"State","States"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        259 : O3EList(122, "InfoDtcList", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Info","Infos"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        260 : O3EList(122, "InfoDtcHistory", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Info","Infos"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        261 : O3EList(122, "ServiceDtcList", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Service","Services"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        262 : O3EList(122, "ServiceDtcHistory", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Service","Services"), O3EDateTime(8, "DateTime"),O3EByteVal(2, "Unknown")] )]),
        265 : O3EList(122, "ErrorDtcList", [O3EByteVal(2, "Count"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Error","Errors"), O3EDateTime(8, "DateTime"), O3EByteVal(2, "Unknown")] )]),
        266 : O3EList(124, "ErrorDtcHistory", [O3EByteVal(2, "Count"), O3EByteVal(2, "GrandTotal"), O3EComplexType(12, "ListEntries",[O3EEnum(2,"Error","Errors"), O3EDateTime(8, "DateTime"), O3EByteVal(2, "Unknown")] )]),         
```

**Reply by M4n197** [reply]
Entschuldige, wer bereits open3e nutzt, kann damit sicherlich bequem an diese Infos gelangen.

In meinem Sonderfall muss ich dafür aber nochmal etwas Zeit investieren, da ich das auf einem ESP32 mittels ESPHome erledige.

**Reply by M4n197** [reply]
@philippoo66 wenn ich mir das so anschaue, wird der String mit der knappen Erläuterung (verständlicherweise) nicht über die Schnittstelle kommuniziert.
Die interne Bezeichnung laut den Enum "Warnings" wäre dann wohl `106: "CondtionsIsolationMonitoring",`? 
Und die zwei unbekannten `O3EByteVal` würden dann evtl. diese ominöse 96 ausspucken... 

Für eine (minimal) weniger kryptische Nachricht hilft dann nur ein Blick auf das Display oder gibt es irgendwo noch eine Art "Katalog" zum Nachschlagen?  In der mir bekannten Anleitung war schon beim Code (`A.106`) Schluss.

**Reply by philippoo66** [reply]
sorry hatte ich grad nicht auf dem Schirm dass du die ESP Geschichte am Laufen hast. 
`Die interne Bezeichnung laut den Enum "Warnings" wäre dann wohl 106: "CondtionsIsolationMonitoring",?`
erscheint mir schlüssig.
`Und die zwei unbekannten O3EByteVal würden dann evtl. diese ominöse 96 ausspucken...`
das würde Sinn machen.

**Reply by philippoo66** [reply]
`gibt es irgendwo noch eine Art "Katalog" zum Nachschlagen?`
die apk Sourcen / der Webcode von Viguide könnten da evtl. weiterführen
